Wikipedia/Code/WMFDatabaseHouseKeeper.swift

```diff
@@ -132,7 +133,19 @@ import Foundation
             let referencedKeysPredicate = NSPredicate(format: "!(key IN %@)", referencedArticleKeys)
             articlesToDeletePredicate = NSCompoundPredicate(andPredicateWithSubpredicates:[articlesToDeletePredicate,referencedKeysPredicate])
         }
-        
+
+        let oldNavigationStackArticlesToDelete: NSPredicate?
+        if let preservedArticleKeys = navigationStateController.allPreservedArticleKeys(in: moc) {
```

Comment:
It seems like `preservedArticleKeys` could be added to `referencedArticleKeys` above to eliminate the need for this additional predicate

Reply:
`viewedDate` won't be `NULL` for the articles from the stack

Reply:
That should be fine. We only want to delete an article if the `viewedDate` is `NULL` AND it's not referenced by the navigation stack or the explore feed. If it's the case now that all articles returned by `allPreservedArticleKeys` have `viewedDate` as non-null, they're already covered by this predicate and we could remove the consideration entirely. Including `allPreservedArticleKeys` keys in  `referencedArticleKeys` as with 9a9c6b would handle any future state when articles are on the stack with a null `viewedDate` - maybe some other view that has an article that isn't persisted to history, but I'm fine with removing that as well.

Reply:
So previously, we were not saving any of the articles fetched in Article VC to disk and now we're saving all of them, and we want to keep most of them? I was thinking that since we're saving all the articles, we would want to be more aggressive about removing them (in the nav stack case)

Reply:
OK - I see where I misunderstood. Currently deletion from Core Data and from disk are linked because only saved articles are stored on disk and need to be deleted from disk. In the change you made, it deletes older articles from Core Data in order to delete them from disk. However, deleting them from Core Data would also remove them from history. It seems like there needs to be an additional consideration to remove these articles from disk while keeping them in Core Data. Do I have that right?

Reply:
Yup that's it 👍 I missed the history deletion part. 😬  So would it make sense to append the urls of the articles where

`"viewedDate != NULL && savedDate == NULL && placesSortOrder == 0 && isExcludedFromFeed == FALSE"` AND `key NOT IN the union of referenced keys & preserved keys`

without deleting them in Core Data

or we could remove the `viewedDate` condition from the predicate and then guard against `viewedDate == nil` before core data deletion

Reply:
I did an initial implementation and the problem I'm seeing is that since the Core Data state remains the same for articles not in the nav stack but still in history, the same set gets added to the disk removal queue every time the app quits. This would be problematic for users with a lot of articles in history as the app would have to iterate through and try to delete all of them every time the app exits. We could use `isDownloaded` to signal whether or not the article is cached and set it to false when returning it with the list of URLs. Currently `isDownloaded` is only used by the saved articles fetcher and includes images so it's set to false to trigger a re-download of images when a new revision is fetched (maybe this is the stuck downloading bug?). We could work around this by adding a separate method to trigger an image update. The other caveat is that this is all going to be re-worked for `mobile-html` so I'm not sure how much effort it's worth now

Reply:
It seems like a separate `isCached` property is clearer, keeping the `isDownloaded` property for the saved articles fetcher. I tried this approach in b254d22